modules/dnn/src/layers/eltwise_layer.cpp

```diff
@@ -137,194 +204,307 @@ class EltwiseLayerImpl CV_FINAL : public EltwiseLayer
+        EltwiseLayerImpl& self;
         std::vector<const Mat*> srcs;
+        std::vector<int> srcNumChannels;
         int nsrcs;
         Mat* dst;
         std::vector<float> coeffs;
-        EltwiseOp op;
         int nstripes;
         const ActivationLayer* activ;
         int channels;
         size_t planeSize;
 
-        EltwiseInvoker() : nsrcs(0), dst(0), op(PROD), nstripes(0), activ(0), channels(0), planeSize(0)  {}
+        EltwiseInvoker(EltwiseLayerImpl& self_)
+            : self(self_)
+            , nsrcs(0), dst(0), nstripes(0), activ(0), channels(0)
+            , planeSize(0)
+        {}
 
-        static void run(const Mat* srcs, int nsrcs, Mat& dst,
-                        const std::vector<float>& coeffs, EltwiseOp op,
-                        const ActivationLayer* activ, int nstripes)
+    public:
+        static void run(EltwiseLayerImpl& self,
+                        const Mat* srcs, int nsrcs, Mat& dst,
+                        int nstripes)
         {
+            const EltwiseOp op = self.op;
             CV_Check(dst.dims, 1 < dst.dims && dst.dims <= 5, ""); CV_CheckTypeEQ(dst.type(), CV_32FC1, ""); CV_Assert(dst.isContinuous());
-            CV_Assert(coeffs.empty() || coeffs.size() == (size_t)nsrcs);
+            CV_Assert(self.coeffs.empty() || self.coeffs.size() == (size_t)nsrcs);
+            CV_CheckGE(nsrcs, 2, "");
+
+            CV_Assert(self.outputChannels == dst.size[1]);
 
-            EltwiseInvoker p;
+            EltwiseInvoker p(self);
             p.srcs.resize(nsrcs);
-            p.coeffs = coeffs;
+            p.srcNumChannels.resize(nsrcs);
+            p.coeffs = self.coeffs;  // can be sorted
+
+            bool sortInputs = false;
             for( int i = 0; i < nsrcs; i++ )
             {
-                p.srcs[i] = srcs + i;
-                CV_Assert(srcs[i].type() == dst.type() &&
-                          srcs[i].isContinuous());
-                // Sort srcs and coefficients in the order by number of channels
-                for( int j = i; j >= 1 && p.srcs[j - 1]->size[1] < p.srcs[j]->size[1]; j-- )
+                p.srcs[i] = &srcs[i];
+                CV_CheckEQ(srcs[i].dims, dst.dims, "");
+                CV_Assert(srcs[i].isContinuous());
+                CV_Assert(srcs[i].type() == dst.type());
+                p.srcNumChannels[i] = (srcs[i].dims >= 4) ? srcs[i].size[1] : 1;
+
+                if (!self.variableChannels)
                 {
-                    std::swap(p.srcs[j - 1], p.srcs[j]);
-                    if (!p.coeffs.empty())
-                        std::swap(p.coeffs[j - 1], p.coeffs[j]);
+                    CV_Assert(srcs[i].size == dst.size);
+                }
+                else if (self.channelsMode == ELTWISE_CHANNNELS_INPUT_0)
+                {
+                    if (i == 0)
+                        CV_Assert(srcs[0].size == dst.size);
+                    CV_Assert(self.outputChannels >= p.srcNumChannels[i]);
+                }
```

Comment:
It seems to me that we need to sort the inputs here as well: 2, 4, 3 channels, in example

Reply:
BTW, processing doesn't require sorted inputs. Result is correct anyway, but may be less efficient.

Updated code.